File: netlify/netlify/functions/get-d-values.py
import json
import os
import boto3
from decimal import Decimal
import datetime

 # Initialize DynamoDB client
dynamodb = boto3.resource(
    "dynamodb",
    region_name=os.environ.get("AWS_REGION"),
    aws_access_key_id=os.environ.get("AWS_ACCESS_KEY_ID"),
    aws_secret_access_key=os.environ.get("AWS_SECRET_ACCESS_KEY"),
)
table_name = os.environ.get("DYNAMODB_TABLE_NAME")
table = dynamodb.Table(table_name)
dynamodb = boto3.resource(
    "dynamodb",
    region_name=os.environ.get("AWS_REGION"),
    aws_access_key_id=os.environ.get("AWS_ACCESS_KEY_ID"),
    aws_secret_access_key=os.environ.get("AWS_SECRET_ACCESS_KEY"),
)
table_name = os.environ.get("DYNAMODB_TABLE_NAME")
table = dynamodb.Table(table_name)


class DecimalEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, Decimal):
            return float(o)
        return json.JSONEncoder.default(self, o)


def handler(event, context):
    try:
        if event["httpMethod"] != "GET":
            return {
                "statusCode": 405,
                "headers": {"Access-Control-Allow-Origin": "*"},
                "body": json.dumps({"error": "Method Not Allowed"}),
            }

        query_params = event.get("queryStringParameters", {})
        start_date_str = query_params.get("startDate")
        end_date_str = query_params.get("endDate")

        # If date parameters are not provided, get data for the last 7 days by default
        if not start_date_str or not end_date_str:
            end_dt = datetime.datetime.now(datetime.timezone.utc)
            start_dt = end_dt - datetime.timedelta(days=7)
            start_date_iso = start_dt.isoformat(timespec="seconds") + "Z"
            end_date_iso = end_dt.isoformat(timespec="seconds") + "Z"
        else:
            # Convert date string from frontend to ISO format (e.g., "YYYY-MM-DD" -> "YYYY-MM-DDT00:00:00Z")
            start_date_iso = start_date_str + "T00:00:00Z"
            end_date_iso = end_date_str + "T23:59:59Z"

        # Execute DynamoDB Query (specify range with partition key and sort key)
        response = table.query(
            KeyConditionExpression=boto3.dynamodb.conditions.Key("recordType").eq(
                "D_VALUE_RECORD"
            )
            & boto3.dynamodb.conditions.Key("timestamp").between(
                start_date_iso, end_date_iso
            )
        )
        items = response["Items"]

        formatted_data = []
        for item in items:
            formatted_data.append(
                {"timestamp": item.get("timestamp"), "d_value": item.get("d_value")}
            )

        formatted_data.sort(key=lambda x: x.get("timestamp", ""))

        return {
            "statusCode": 200,
            "headers": {
                "Content-Type": "application/json",
                "Access-Control-Allow-Origin": "*",  # TODO: Restrict to frontend domain
                "Access-Control-Allow-Methods": "GET",
                "Access-Control-Allow-Headers": "Content-Type",
            },
            "body": json.dumps(formatted_data, cls=DecimalEncoder),
        }

    except Exception as e:
        print(f"Error fetching data from DynamoDB: {e}")
        return {
            "statusCode": 500,
            "headers": {
                "Content-Type": "application/json",
                "Access-Control-Allow-Origin": "*",
            },
            "body": json.dumps({"error": str(e)}),
        }

File: netlify/netlify/functions/get-d-values.ts
import { Handler, APIGatewayEvent, Context } from 'aws-lambda';
import AWS from 'aws-sdk';

// AWS SDK configuration (get credentials from Netlify environment variables)
AWS.config.update({
    region: process.env.NETLIFY_AWS_REGION,
    accessKeyId: process.env.NETLIFY_AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.NETLIFY_AWS_SECRET_ACCESS_KEY,
});

const docClient = new AWS.DynamoDB.DocumentClient();
const tableName = process.env.DYNAMODB_TABLE_NAME;

export const handler: Handler = async (event: APIGatewayEvent, context: Context) => {
    try {
        if (event.httpMethod !== 'GET') {
            return {
                statusCode: 405,
                headers: { 'Access-Control-Allow-Origin': '*' },
                body: JSON.stringify({ error: 'Method Not Allowed' }),
            };
        }

        const queryParams = event.queryStringParameters || {};
        let startDateStr = queryParams.startDate;
        let endDateStr = queryParams.endDate;

        let startDateIso: string;
        let endDateIso: string;

        // If date parameters are not provided, get data for the last 7 days by default
        if (!startDateStr || !endDateStr) {
            const now = new Date();
            const endDt = new Date(now.toISOString()); // UTC
            const startDt = new Date(endDt.getTime() - 7 * 24 * 60 * 60 * 1000); // 7 days ago

            startDateIso = startDt.toISOString().split('.')[0] + 'Z';
            endDateIso = endDt.toISOString().split('.')[0] + 'Z';
        } else {
            // Convert date string from frontend (YYYY-MM-DD) to ISO format
            startDateIso = `${startDateStr}T00:00:00Z`;
            endDateIso = `${endDateStr}T23:59:59Z`;
        }

        if (!tableName) {
            throw new Error('DynamoDB table name is not configured.');
        }

        const params: AWS.DynamoDB.DocumentClient.QueryInput = {
            TableName: tableName,
            KeyConditionExpression: 'recordType = :rt AND #ts BETWEEN :startTs AND :endTs',
            ExpressionAttributeNames: {
                '#ts': 'timestamp', // Use alias for 'timestamp' in case it is a reserved word
            },
            ExpressionAttributeValues: {
                ':rt': 'D_VALUE_RECORD',
                ':startTs': startDateIso,
                ':endTs': endDateIso,
            },
        };

        const result = await docClient.query(params).promise();
        const items = result.Items || [];

        const formattedData = items.map((item) => ({
            timestamp: item.timestamp,
            d_value: item.d_value,
        }));

        // Sort by timestamp (DynamoDB Query sorts by sort key, but do it just in case)
        formattedData.sort((a, b) => {
            if (a.timestamp < b.timestamp) return -1;
            if (a.timestamp > b.timestamp) return 1;
            return 0;
        });

        return {
            statusCode: 200,
            headers: {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*', // In production, restrict to frontend domain
                'Access-Control-Allow-Methods': 'GET',
                'Access-Control-Allow-Headers': 'Content-Type',
            },
            body: JSON.stringify(formattedData),
        };
    } catch (error: any) {
        console.error('Error fetching data from DynamoDB:', error);
        return {
            statusCode: 500,
            headers: {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            body: JSON.stringify({ error: error.message || 'Unknown error' }),
        };
    }
};
